src/liballoc/raw_vec.rs

```diff
@@ -562,6 +522,85 @@ impl<T, A: AllocRef> RawVec<T, A> {
     }
 }
 
+// This was once part of `RawVec::grow`, but pulling it out into a function
+// that's not generic over `T` reduces code size significantly because
+// `RawVec::grow` is often instantiated for many different `T`s.
+fn grow<A>(
+    strategy: Strategy,
+    placement: ReallocPlacement,
+    init: AllocInit,
+    elem_size: usize,
+    elem_align: usize,
+    cap: usize,
+    current_memory: Option<(NonNull<u8>, Layout)>,
+    alloc: &mut A,
+) -> Result<MemoryBlock, TryReserveError>
+where
+    A: AllocRef,
+{
+    if elem_size == 0 {
+        // Since we return a capacity of `usize::MAX` when `elem_size` is
+        // 0, getting to here necessarily means the `RawVec` is overfull.
+        return Err(CapacityOverflow);
+    }
```

Comment:
Have you considered moving this check back into the generic method where it can be optimized away?

Reply:
Good suggestion, I will try that.